internal/scan/scanner.go

```diff
@@ -0,0 +1,271 @@
+package scan
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/stellar/go/support/datastore"
+	"github.com/stellar/go/support/log"
+)
+
+// Report represents the aggregated outcome of a full ledger scan.
+// It summarizes all gaps, counts, and the overall ledger range discovered.
+type Report struct {
+	Gaps         []Gap  `json:"gaps,omitempty"` // All missing ledger ranges found across partitions.
+	TotalFound   uint32 `json:"total_found"`    // Total number of ledgers successfully found.
+	TotalMissing uint64 `json:"total_missing"`  // Total number of missing ledgers across all gaps.
+	Min          uint32 `json:"min"`            // Lowest ledger sequence number observed.
+	Max          uint32 `json:"max"`            // Highest ledger sequence number observed.
+}
+
+// Gap represents a contiguous range of missing ledgers detected
+// between existing data ranges.
+type Gap struct {
+	Start uint32 `json:"start"` // First missing ledger in the range.
+	End   uint32 `json:"end"`   // Last missing ledger in the range.
+}
+
+// Partition defines a contiguous range of ledger sequences
+// that should be scanned together by one worker.
+type Partition struct {
```

Comment:
I think partition is a bit overloaded since we also have that as a property of the datastore schema. Since this data structure is used by the workers, I think a better name would be "task" because this represents the work performed by each worker.

Also, all of the structs / functions which are internal to this package should not be exposed. I think this is an example of a struct which should not be public and stay local to the package